mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
     @Override
     public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagUpdateCalculator, MessageRange set) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        return messageIdDAO.retrieveMessages(mailboxId, set)
-                .join()
-                .flatMap(message -> updateFlagsOnMessage(mailbox, flagUpdateCalculator, message))
-                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                    .thenApply(voidValue -> updatedFlags))
-                .map(CompletableFuture::join)
-                .collect(Collectors.toList()) // This collect is here as we need to consume all the stream before returning result
-                .iterator();
+
+        return runUpdate(mailboxId, set, flagUpdateCalculator).iterator();
+    }
+
+    private List<UpdatedFlags> runUpdate(CassandraId mailboxId, MessageRange set, FlagsUpdateCalculator flagsUpdateCalculator) throws MailboxException {
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, set).join();
+
+        FlagsUpdateStageResult globalResult = runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator);
+
+        int retryCount = 0;
+
+        while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
+            retryCount++;
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId,
+                FluentFutureStream.of(
```

Comment:
If I understand your probably-too-smart algorithm, you retry the update on failed flags only, isn't ? I suggest to find a way to write it so that everybody can understand in less than 2 minutes (I needed more than that and not sure I'm right)